src/core/ext/filters/stateful_session/stateful_session_filter.cc

```diff
@@ -158,29 +189,46 @@ ArenaPromise<ServerMetadataHandle> StatefulSessionFilter::MakeCallPromise(
               this, cookie_config->name->c_str(),
               std::string(*cookie_value).c_str());
     }
-    // We have a valid cookie, so add the call attribute to be used by the
-    // xds_override_host LB policy.
-    service_config_call_data->SetCallAttribute(
-        GetContext<Arena>()->New<XdsOverrideHostAttribute>(*cookie_value));
+    std::pair<absl::string_view, absl::string_view> host_cluster =
+        absl::StrSplit(*cookie_value, absl::MaxSplits(';', 1));
+    if (!host_cluster.first.empty() &&
+        OverrideClusterIfNeeded(host_cluster.second, service_config_call_data,
+                                GetContext<Arena>())) {
+      // We have a valid cookie, so add the call attribute to be used by the
+      // xds_override_host LB policy.
+      service_config_call_data->SetCallAttribute(
+          GetContext<Arena>()->New<XdsOverrideHostAttribute>(
+              host_cluster.first));
+    }
+    // Cluster is not yet in use
+  }
+  absl::string_view cluster_name;
+  auto cluster_name_attribute = static_cast<XdsClusterAttribute*>(
+      service_config_call_data->GetCallAttribute(
+          XdsClusterAttribute::TypeName()));
+  if (cluster_name_attribute != nullptr) {
+    cluster_name = cluster_name_attribute->cluster();
   }
   // Intercept server initial metadata.
   call_args.server_initial_metadata->InterceptAndMap(
-      [cookie_config, cookie_value](ServerMetadataHandle md) {
+      [cookie_config, cookie_value, cluster_name](ServerMetadataHandle md) {
```

Comment:
I don't think it's safe to store the `absl::string_view` pointing to the cluster name in this lambda, because that `string_view` points to the storage inside the `ClusterRef` object, and that object may have been destroyed by the time the lambda runs, because the call will drop its ref to the `ClusterRef` as soon as it is committed (i.e., as soon as it's done with all LB picks), which happens before we see server initial metadata.  To fix this, I think we're going to need to allocate a copy of the cluster string on the arena.  See my comment above about moving the cluster string arena allocation into `OverrideClusterIfNeeded()`, which I am suggesting renaming to `GetClusterNameToUse()`.

With that change in place, I think the logic here should look something like this:

```
std::string cookie_value = GetOverrideCookie(
    call_args.client_initial_metadata, *cookie_config->name);
absl::string_view cluster_from_cookie;
// Cookie format is "host;cluster".
std::pair<absl::string_view, absl::string_view> host_cluster =
    absl::StrSplit(cookie_value, absl::MaxSplits(';', 1));
// Set override host attribute.  Allocate the string on the
// arena, so that it has the right lifetime.
if (!host_cluster.first.empty()) {
  char* host_str = GetContext<Arena>()->Alloc(host_cluster.first.size());
  memcpy(host_str, host_cluster.first.data(), host_cluster.first.size());  
  service_config_call_data->SetCallAttribute(
      GetContext<Arena>()->New<XdsOverrideHostAttribute>(
          absl::string_view(host_str, host_cluster.first.size())));
}
// Check if the cluster override is valid, and apply it if necessary.
// Note that cluster_name will point to an arena-allocated string
// that will still be alive when we see the server initial metadata.
// If the cluster name is empty, that means we cannot use a
// cluster override (i.e., the route uses a cluster specifier plugin).
absl::string_view cluster_name =
    GetClusterNameToUse(host_cluster.second, service_config_call_data);
```